### PR TITLE
Re-add mention autocomplete to comments, fixing bugs

### PR DESCRIPTION
**Pull request by aitchiss**
## What type of PR is this? (check all applicable)

- [X] Feature

## Description

**This PR reverts the revert of mention autocomplete in comments, adding some bug fixes.**

After shipping #13061 we realised that a side effect of turning the whole text area into a Reach UI combobox was that navigation within the text area using the up/down arrow keys was no longer functioning. Further investigation showed that Reach UI calls `preventDefault` on these keypresses which is why they weren't registering.

This PR switches up our approach to get around this. The MentionAutocomplete component now creates **two** textareas:

- one vanilla textarea
- one combobox textarea which is initially hidden with `display 'none'`

When a user activates the autosuggest, the vanilla textarea is hidden, and the combobox takes over. When search is exited (either by selection, spacing away, or deleting all text including the `@` sign), the reverse happens. As a text area takes over, focus is shifted so that the user can keep typing without interruption.

This PR also adds a fix for an issue where editing a comment was not showing the original comment pre-populated in the comment box.

## Related Tickets & Documents

https://github.com/forem/rfcs/pull/16
#13061

## QA Instructions, Screenshots, Recordings

The changes here affect all the comment areas mentioned in the original PR:

- comments on posts
- replies to comments on posts
- replies to comments in notifications
- editing of those comments

Steps to test post comments:

- Go to a post
- Click to add a comment
- Start typing, including some @mentions (you might want to add some users with similar usernames to your local DB to be able to see fuller lists appearing in the dropdown)
- Experiment with selecting options with keyboard (up/down arrows + Enter) and mouse
- Experiment with deleting portions of the mention, ignoring the options and keeping typing, etc - try to break it 😄
- Repeat as above with replies to comments on a post

Steps to test post comment replies in notifications:

- Send a notification to your current user in the console:
```
user = User.find_by(username: "admin_mcadmin")
other_user = User.first
article = user.articles.last
comment = article.comments.create!(user: other_user, body_markdown: "Test something")
Notification.send_new_comment_notifications_without_delay(comment)
```
- Refresh the page and go to the notification
- Click reply and repeat the above steps

In all cases, experiment with editing your comment after saving

https://user-images.githubusercontent.com/20773163/112485632-0cc62600-8d73-11eb-8127-5e6d39450890.mp4

### UI accessibility concerns?

As previously noted in the autocomplete work, there are definitely accessibility considerations. This reworked approach has a nice accessibility side effect of only triggering the combobox when the `@mention` has begun. This means when a user types the `@` sign, the screen reader announces that we're now in a combobox. This is similar to the behaviour observed in e.g. GitHub's comments. The previously tested combobox functionality continues to work for screen reader users.

tl;dr I think this is actually a bit better for accessibility than the original implementation.

A screen recording of me testing this out in Safari with VoiceOver:

https://user-images.githubusercontent.com/20773163/112485270-a93bf880-8d72-11eb-8134-141fc0ee2f4a.mp4

## Added tests?

- [X] Yes: Updated the original tests and added a new test for pre-population of comments being edited

## [Forem core team only] How will this change be communicated?

- [X] This change does not need to be communicated, and this is why not: we haven't yet shared the autocomplete functionality in a changelog

## [optional] What gif best describes this PR or how it makes you feel?

dog rolls around saying work work

### CODE REVIEW

File: app/assets/stylesheets/components/autocomplete.scss [hunk 1]
```diff
@@ -2,6 +2,11 @@
 @import '@reach/combobox/styles';
 
 .crayons-autocomplete {
+  // Override of the default hidden class to make sure it takes priority over any display style attached to the 'replaceElement' textarea
+  .hidden {
```
Comment (Link2Twenty) [inline app/assets/stylesheets/components/autocomplete.scss:6]: Is this for when the element has both `crayons-autocomplete` and `hidden` as classes?

If so I think you should include the ampersand 

```css
  &.hidden {
```

Comment (aitchiss) [reply app/assets/stylesheets/components/autocomplete.scss:6]: It's for children of the `crayons-autocomplete` class that have the 'hidden' class 🙂

File: app/assets/stylesheets/components/autocomplete.scss [hunk 2]
```diff
@@ -2,6 +2,11 @@
...
+    display: none !important;
```
Comment (Link2Twenty) [inline app/assets/stylesheets/components/autocomplete.scss:7]: Are the styles you're trying to overwrite on the elements themselves? If not can we just make the rule more specific? It just feels a shame to use important 😅

File: app/assets/stylesheets/components/autocomplete.scss (hunk 1 above)
Comment (Link2Twenty) [reply app/assets/stylesheets/components/autocomplete.scss:6]: TIL: that the ampersand is optional for children 👍 
I'd have expected

```scss
.crayons-autocomplete {
  & .hidden {
  
  }
}
```

who knew it was optional 😁

**Review by rhymes — APPROVED** [review]
It works great, awesome work @aitchiss  😭 

Love the approach of having two modes! This should also help when/if we add emoji autocompletion and other things to the textarea, it'll just be another mode :D

File: app/assets/stylesheets/components/autocomplete.scss (hunk 2 above)
Comment (aitchiss) [reply app/assets/stylesheets/components/autocomplete.scss:7]: I know what you mean it always feels so wrong to use `important` 😄 

In the current use case in comments, I think we could actually omit the `important`, but I'm adding it as future-proofing. The way this autocomplete component works, is we copy all styles from some other element (e.g. the initially loaded comment box), and apply them to two inputs, one of which needs to be hidden at all times. The original element is then removed completely.

I feel like in this case, `important` is a good fit - it makes sure that no matter how specific a style comes from the original element, we'll definitely never have a situation where one of those hidden elements is visible.

It's a bit like in our `reset.scss` where we make sure animations are removed for users with `prefers-reduced-motion`, in that it's a situation where regardless of any previous style, wherever this component is used, we want to override it.

File: app/assets/stylesheets/components/autocomplete.scss (hunk 1 above)
Comment (aitchiss) [reply app/assets/stylesheets/components/autocomplete.scss:6]: yeah I love the nested styling in scss 😍

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx [hunk 3]
```diff
@@ -15,30 +15,51 @@ const MIN_SEARCH_CHARACTERS = 2;
...
+    });
   });
 
-  // Make sure all styles are copied to the autocomplete textarea
-  newNode.style.cssText = document.defaultView.getComputedStyle(
+  // Make sure all styles are copied to the autocomplete & plain textareas
+  const cssText = document.defaultView.getComputedStyle(
     originalNodeToReplace,
     '',
   ).cssText;
```
Comment (nickytonline) [inline app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:47]: Refactor (blocking): Use destructuring.

```suggestion
  const { cssText } = document.defaultView.getComputedStyle(
    originalNodeToReplace,
    '',
  );
```

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx [hunk 4]
```diff
@@ -15,30 +15,51 @@ const MIN_SEARCH_CHARACTERS = 2;
...
 
-  // We need to manually remove the element, as Preact's diffing algorithm won't replace it in render
+  newNodes.forEach((node) => {
```
Comment (nickytonline) [inline app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:49]: Refactor (blocking): Instead of two `forEach` loops in the `Object.keys.forEach` loop, consolidate them as one `forEach` and set the `cssText` before the loop starts.

**Review by nickytonline — CHANGES_REQUESTED** [review]
I tested this locally and all seems to be working well. Some comments and request changes. Looking forward to seeing this back in prod. 😎

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx [hunk 5]
```diff
@@ -203,15 +249,25 @@ export const MentionAutocompleteTextArea = ({
...
   };
 
   useLayoutEffect(() => {
-    if (inputRef.current) {
-      // Replace the whole textarea passed in props with the autocomplete textarea
-      replaceTextArea(replaceElement, inputRef.current);
+    const { current: comboboxTextArea } = comboboxRef;
+    const { current: plainTextArea } = plainTextAreaRef;
+
+    if (comboboxTextArea && plainTextArea) {
```
[the remark this answers is not in the document]
Comment (aitchiss) [reply app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:265]: Yeah I thought a little bit about this and ultimately decided we should only introduce it when we need it, otherwise it's extra code/tests we're not using yet (I think we would also need to e.g. check that the tag type is either input or textarea)

File: cypress/integration/articleFlows/commentOnArticle.spec.js [hunk 6]
```diff
@@ -0,0 +1,268 @@
+describe('Comment on articles', () => {
```
[the remark this answers is not in the document]
Comment (aitchiss) [reply cypress/integration/articleFlows/commentOnArticle.spec.js:1]: I thought about this, but because we now switch from plain textarea to combobox textarea, I think in each test asserting we're in the correct textarea is sufficient? Otherwise we're just testing standard textarea functionality 🙃 

It's a bit of a pity the Cypress tests in this case have to be quite coupled to the implementation - at each step we need to get a new handle to the currently presented textarea

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx (hunk 4 above)
Comment (aitchiss) [reply app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:49]: There's only one `forEach` inside the `Object.keys.forEach` loop 🤔 

Do you mean inside this `forEach` on line 49 we bring inside the `Object.keys.forEach`? It would mean looping over more iterations altogether (since there's always going to be more attributes than newNodes). Maybe I'm misunderstanding though and need more coffee 😆

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx (hunk 3 above)
Comment (aitchiss) [reply app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:47]: I'm gonna make myself a todo to add this to the linter rules 🙂

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx (hunk 4 above)
Comment (Link2Twenty) [reply app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:49]: Could you do the `Object.keys(attributes).forEach((attributeKey)` loop inside the first `newNodes.forEach((node)` loop? Then you could move the styles loop up there too and reduce iterations.

```js
// Make sure all styles are copied to the autocomplete & plain textareas
const cssText = document.defaultView.getComputedStyle(
  originalNodeToReplace,
  '',
).cssText;

const attributes = originalNodeToReplace.attributes;
newNodes.forEach((node) => {
  Object.keys(attributes).forEach((attributeKey) => {
    node.setAttribute(
      attributes[attributeKey].name,
      attributes[attributeKey].value,
    );
  });
  
  node.style.cssText = cssText;
  // Make sure no transition replays when the new textareas are mounted
  node.style.transition = 'none';
  // Copy any initial value
  node.value = originalNodeToReplace.value;
});
```

Comment (aitchiss) [reply app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:49]: I think that's what @nickytonline was suggesting - thanks @Link2Twenty - my brain hadn't quite woken up this morning 😆

Comment (nickytonline) [reply app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:49]: > There's only one `forEach` inside the `Object.keys.forEach` loop 🤔 
> 
> 
> 
> Do you mean inside this `forEach` on line 49 we bring inside the `Object.keys.forEach`? It would mean looping over more iterations altogether (since there's always going to be more attributes than newNodes). Maybe I'm misunderstanding though and need more coffee 😆 

Maybe I need more coffee. 😎 I'll double check this. Maybe I was seeing double from the git diff. 😅

Comment (aitchiss) [reply app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx]: I've switched it up so that we set the attributes inside the `newNodes.forEach...`. My brain is definitely engaging Friday-mode already 🤣

File: app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx [hunk 7]
```diff
@@ -15,30 +15,49 @@ const MIN_SEARCH_CHARACTERS = 2;
...
+const replaceTextArea = ({
+  originalNodeToReplace,
+  plainTextArea,
+  comboboxTextArea,
+}) => {
+  const newNodes = [plainTextArea, comboboxTextArea];
 
-  // Make sure all styles are copied to the autocomplete textarea
-  newNode.style.cssText = document.defaultView.getComputedStyle(
+  const attributes = originalNodeToReplace.attributes;
```
Comment (nickytonline) [inline app/javascript/crayons/MentionAutocompleteTextArea/MentionAutocompleteTextArea.jsx:32]: Refactor (non-blocking): Prefer destructuring. I know you mentioned adding the eslint rule, so we should be good in future PRs. 😎

**Review by nickytonline — APPROVED** [review]
This looks great @aitchiss! One non-blocking suggestion, so this is good to go. 🚢 Looking forward to seeing and using this in production. 🌟
